### PR TITLE
[warm-reboot][sad] enhance test to cover different port types and ope…

### PR DESCRIPTION
**Pull request by stepanblyschak**
…r state in sad scenarios

Signed-off-by: Stepan Blyschak <stepanb@nvidia.com>

### Description of PR

Summary: Enhance warm reboot sad path test cases.
Fixes ```https://github.com/Azure/sonic-mgmt/issues/3683```

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

### Back port request
- [ ] 201911

#### What is the motivation for this PR?

The motivation is to fill the test gap to test different ports admin and operationally down before warm-reboot and check their status after warm-reboot. Sanity will make sure the ports changed are back up.

#### How did you do it?

- For port down cases implemented a special logic for port selection based on port physical properties.
- For port down cases implemented also port operational state change.
- As of second enhancement in the list there was a neccessary refactoring made, few sad operations were taken out from ptf script and implemented in pytest. Old sad path wasn't changed to support ansible compatiblity. This is partial refactoring only for enabling us to implement first two items from this list. Other sad operations need to be taken to pytest as well in the future.

#### How did you verify/test it?

Running warm_reboot_multi_sad.

### CODE REVIEW

File: tests/common/helpers/sad_path.py [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+        """ Verify handler that runs after warm-reboot completes.
+        Checks sad path operataions result after warm-reboot. """
+        raise NotImplementedError
+
+    def revert(self):
+        """ Revert changes done in setup. """
+        raise NotImplementedError
+
+
+class PhyPropsPortSelector:
```
Comment (stepanblyschak) [inline tests/common/helpers/sad_path.py]: Add ISelector interface.

File: tests/platform_tests/test_advanced_reboot.py [hunk 2]
```diff
@@ -94,34 +106,35 @@ def test_warm_reboot_sad(request, get_advanced_reboot, advanceboot_neighbor_rest
...
     prebootList = [
-        'neigh_bgp_down:2',             # Shutdown single BGP session on 2 remote devices (VMs) before reboot DUT
-        'dut_bgp_down:3',               # Shutdown 3 BGP sessions on DUT brefore rebooting it
-        'dut_lag_down:2',               # Shutdown 2 LAG sessions on DUT brefore rebooting it
-        'neigh_lag_down:3',             # Shutdown 1 LAG session on 3 remote devices (VMs) before reboot DUT
-        'dut_lag_member_down:3:1',      # Shutdown 1 LAG member of 3 LAG sessions corresponding to 3 remote devices (VM)
-                                        # on DUT
-        'neigh_lag_member_down:2:1',    # Shutdown 1 LAG member of 2 LAG sessions on 2 remote devices (VM) (1 each)
-        'vlan_port_down:4',
+         'neigh_bgp_down:2',             # Shutdown single BGP session on 2 remote devices (VMs) before reboot DUT
```
Comment (stepanblyschak) [inline tests/platform_tests/test_advanced_reboot.py]: bad formatting

File: tests/common/helpers/sad_path.py (hunk 1 above)
Comment (stepanblyschak) [reply tests/common/helpers/sad_path.py]: Added

File: tests/platform_tests/test_advanced_reboot.py (hunk 2 above)
Comment (stepanblyschak) [reply tests/platform_tests/test_advanced_reboot.py]: Fixed

**Review by nazariig — APPROVED** [review]

**Pull request closed without merging**